File: 74-search-a-2d-matrix/search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int rows = matrix.size()-1;
        int cols = matrix[0].size()-1;
        int low = 0, high = rows;
        
        while(low <= high){
            int mid_row = (low+high)/2;
            
            if(target >= matrix[mid_row][0] && target <= matrix[mid_row][cols]){ 
                int l = 0, r = cols;
                while(l<=r){
                    int mid = (l+r)/2;
                    if(matrix[mid_row][mid] == target) return true;
                    else if(matrix[mid_row][mid] < target) l = mid+1;
                    else r = mid - 1;
                }
                return false;
            }
            else if(matrix[mid_row][cols] < target) low = mid_row + 1;
            else high = mid_row - 1;
        }
        return false;
    }
};

File: 74-search-a-2d-matrix/search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        low, high = 0, len(matrix) - 1

        while(low<=high):
            midRow = (low+high)//2
            if(target >= matrix[midRow][0] and target <= matrix[midRow][-1]):
                l, r = 0, len(matrix[0]) - 1              
                while(l<=r):
                    mid = (l+r)//2
                    if matrix[midRow][mid] == target:
                        return True
                    elif matrix[midRow][mid] > target: r = mid - 1
                    else: l = mid + 1

                return False

            elif target > matrix[midRow][-1]: low = midRow + 1
            else: high = midRow - 1
        
        return False
